File: lib/native/features/download/data/backend_service/api/service_result.dart
class ServiceResult<T> {
  bool isSuccess;
  Object? error;
  T? data;

  ServiceResult.success(this.data) : isSuccess = true;
  ServiceResult.error(this.error) : isSuccess = false;

  String getErrorMessage() {
    return error.toString();
  }

  ServiceResult(this.isSuccess, this.error, this.data);

  Map<String, dynamic> toJson() => {
        'isSuccess': isSuccess,
        'error': error,
        'data': data,
      };

  factory ServiceResult.fromJson(Map<String, dynamic> json) => ServiceResult(
        json['isSuccess'],
        json['error'],
        json['data'],
      );
}

File: android/app/src/main/kotlin/com/example/perfect_corp_homework/api/ServiceResult.kt
package com.example.perfect_corp_homework.api

class ServiceResult<T> {

    constructor(error: Exception) {
        this.isSuccess = false
        this.error = error
        this.data = null
    }

    constructor(data: T) {
        this.isSuccess = true
        this.error = null
        this.data = data
    }

    val isSuccess: Boolean
    val error: Exception?
    val data: T?

}